pkg/progressreader/progressreader_test.go

```diff
@@ -0,0 +1,80 @@
+package progressreader
+
+import (
+	"bufio"
+	"bytes"
+	"io"
+	"io/ioutil"
+	"testing"
+
+	"github.com/docker/docker/pkg/streamformatter"
+)
+
+func TestOutputOnPrematureClose(t *testing.T) {
+	var outBuf bytes.Buffer
+	content := []byte("TESTING")
+	reader := ioutil.NopCloser(bytes.NewReader(content))
+	writer := bufio.NewWriter(&outBuf)
+
+	prCfg := Config{
+		In:        reader,
+		Out:       writer,
+		Formatter: streamformatter.NewStreamFormatter(),
+		Size:      len(content),
+		NewLines:  true,
+		ID:        "Test",
+		Action:    "Read",
+	}
+	pr := New(prCfg)
+
+	part := make([]byte, 4, 4)
+	_, err := io.ReadFull(pr, part)
+	if err != nil {
+		pr.Close()
+		t.Fatal(err)
+	}
+
+	writer.Flush()
+	tlen := outBuf.Len()
+	pr.Close()
+	writer.Flush()
+	if outBuf.Len() == tlen {
+		t.Fatalf("Expected some output when closing prematurely")
+	}
+}
+
+func TestCompleteSilently(t *testing.T) {
+	var outBuf bytes.Buffer
+	content := []byte("TESTING")
+	reader := ioutil.NopCloser(bytes.NewReader(content))
+	writer := bufio.NewWriter(&outBuf)
+
+	prCfg := Config{
+		In:        reader,
+		Out:       writer,
+		Formatter: streamformatter.NewStreamFormatter(),
+		Size:      len(content),
+		NewLines:  true,
+		ID:        "Test",
+		Action:    "Read",
+	}
+	pr := New(prCfg)
+
+	out, err := ioutil.ReadAll(pr)
+	if err != nil {
+		pr.Close()
+		t.Fatal(err)
+	}
+	if string(out) != "TESTING" {
+		pr.Close()
+		t.Fatalf("Unexpected output %q from reader", string(out))
+	}
+
+	writer.Flush()
+	tlen := outBuf.Len()
+	pr.Close()
```

Comment:
@runcom given that `outBuf` grows automatically, the only error I can conceive happening here would be an 'out of memory' if progressreader outputs too much. Still, you're right, it's better to be disciplined when fiddling with byte arrays :)